graphql/schemabuilder/reflect.go

```diff
@@ -498,81 +527,137 @@ func (sb *schemaBuilder) buildFunction(typ reflect.Type, m *method) (*graphql.Fi
-			// Set up function arguments.
-			in := make([]reflect.Value, 0, funcType.NumIn())
+	return result, nil
 
-			if hasContext {
-				in = append(in, reflect.ValueOf(ctx))
-			}
+}
 
-			// Set up source.
-			if hasSource {
-				sourceValue := reflect.ValueOf(source)
-				ptrSource := sourceValue.Kind() == reflect.Ptr
-				switch {
-				case ptrSource && !ptrFunc:
-					in = append(in, sourceValue.Elem())
-				case !ptrSource && ptrFunc:
-					copyPtr := reflect.New(typ)
-					copyPtr.Elem().Set(sourceValue)
-					in = append(in, copyPtr)
-				default:
-					in = append(in, sourceValue)
-				}
-			}
+type funcContext struct {
+	hasContext      bool
+	hasSource       bool
+	hasArgs         bool
+	hasSelectionSet bool
+	hasRet          bool
+	hasError        bool
 
-			// Set up other arguments.
-			if hasArgs {
-				in = append(in, reflect.ValueOf(args))
-			}
-			if hasSelectionSet {
-				in = append(in, reflect.ValueOf(selectionSet))
-			}
+	funcType     reflect.Type
+	isPtrFunc    bool
+	typ          reflect.Type
+	selectionSet *graphql.SelectionSet
+}
+
+func (funcCtx *funcContext) setUpResolve(source interface{}, args interface{}, ctx context.Context) []reflect.Value {
+
+	in := make([]reflect.Value, 0, funcCtx.funcType.NumIn())
+	if funcCtx.hasContext {
+		in = append(in, reflect.ValueOf(ctx))
+	}
+
+	// Set up source.
+	if funcCtx.hasSource {
+		sourceValue := reflect.ValueOf(source)
+		ptrSource := sourceValue.Kind() == reflect.Ptr
+		switch {
+		case ptrSource && !funcCtx.isPtrFunc:
+			in = append(in, sourceValue.Elem())
+		case !ptrSource && funcCtx.isPtrFunc:
+			copyPtr := reflect.New(funcCtx.typ)
+			copyPtr.Elem().Set(sourceValue)
+			in = append(in, copyPtr)
+		default:
+			in = append(in, sourceValue)
+		}
+	}
+
+	// Set up other arguments.
+	if funcCtx.hasArgs {
+		in = append(in, reflect.ValueOf(args))
+	}
+	if funcCtx.hasSelectionSet {
+		in = append(in, reflect.ValueOf(funcCtx.selectionSet))
+	}
+
+	return in
+
+}
+
+func (sb *schemaBuilder) buildFunction(typ reflect.Type, m *method) (*graphql.Field, error) {
+
+	funcCtx := &funcContext{typ: typ}
+
+	fun, in, err := funcCtx.getFuncValAndType(m)
+	if err != nil {
+		return nil, err
+	}
+
+	in = funcCtx.consumeContextAndSource(in)
+
+	argParser, argType, in, err := sb.getArgParserAndTyp(in)
+	if err != nil {
+		return nil, err
+	}
+	funcCtx.hasArgs = argParser != nil
+
+	in = funcCtx.consumeSelectionSet(in)
+
+	// We have succeeded if no arguments remain.
+	if len(in) != 0 {
+		return nil, fmt.Errorf("%s arguments should be [context][, [*]%s][, args][, selectionSet]", funcCtx.funcType, typ)
+	}
+
+	// Parse return values. The first return value must be the actual value, and
+	// the second value can optionally be an error.
+	err = funcCtx.consumeRetAndError(m)
+	if err != nil {
+		return nil, err
+	}
 
+	retType, err := sb.getRetType(funcCtx.hasRet, funcCtx.funcType, m)
+	if err != nil {
+		return nil, err
+	}
+
+	args, err := funcCtx.getArgsMap(argType)
```

Comment:
what about `argsTypeMap`? `getArgsRetType`? makes it more clear that it is for the graphql typing

Reply:
argstypemaps sounds nice